homeassistant/components/stream/hls.py

```diff
@@ -64,55 +78,184 @@ class HlsPlaylistView(StreamView):
     cors_allowed = True
 
     @staticmethod
-    def render_preamble(track):
-        """Render preamble."""
-        return [
-            "#EXT-X-VERSION:7",
-            f"#EXT-X-TARGETDURATION:{track.target_duration}",
-            '#EXT-X-MAP:URI="init.mp4"',
-        ]
+    def render(track, ll_hls):
+        """Render HLS playlist file."""
+        # NUM_PLAYLIST_SEGMENTS+1 because most recent is probably not yet complete
+        segments = list(track.get_segment())[-(NUM_PLAYLIST_SEGMENTS + 1) :]
 
-    @staticmethod
-    def render_playlist(track):
-        """Render playlist."""
-        segments = list(track.get_segment())[-NUM_PLAYLIST_SEGMENTS:]
+        # Create a copy of the last segment with the parts frozen for rendering playlist
+        segments[-1] = copy(segments[-1])
+        segments[-1].parts = segments[-1].parts.copy()
+
+        # Remove first segment if the last segment is actually complete
+        if segments[-1].duration:
+            segments = segments[-NUM_PLAYLIST_SEGMENTS:]
 
-        if not segments:
-            return []
+        target_duration = round(max(s.duration for s in segments if s.duration))
 
         playlist = [
+            "#EXTM3U",
+            "#EXT-X-VERSION:6",
+            "#EXT-X-INDEPENDENT-SEGMENTS",
+            '#EXT-X-MAP:URI="init.mp4"',
+            f"#EXT-X-TARGETDURATION:{target_duration}",
             f"#EXT-X-MEDIA-SEQUENCE:{segments[0].sequence}",
             f"#EXT-X-DISCONTINUITY-SEQUENCE:{segments[0].stream_id}",
+            "#EXT-X-PROGRAM-DATE-TIME:"
+            + segments[0].start_time.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3]
+            + "Z",
         ]
 
-        last_stream_id = segments[0].stream_id
-        for segment in segments:
-            if last_stream_id != segment.stream_id:
-                playlist.append("#EXT-X-DISCONTINUITY")
+        if ll_hls:
+            part_duration = float(
+                max(
+                    (p.duration for s in segments for p in s.parts),
+                    default=PART_DURATION,
+                )
+            )
             playlist.extend(
                 [
-                    f"#EXTINF:{float(segment.duration):.04f},",
-                    f"./segment/{segment.sequence}.m4s",
+                    f"#EXT-X-PART-INF:PART-TARGET={part_duration:.3f}",
+                    f"#EXT-X-SERVER-CONTROL:CAN-BLOCK-RELOAD=YES,PART-HOLD-BACK={2*part_duration:.3f}",
+                    f"#EXT-X-START:TIME-OFFSET=-{3*part_duration:.3f},PRECISE=YES",
                 ]
             )
+        else:
+            # Since our window doesn't have many segments, we don't want to start
+            # at the beginning or we risk a behind live window exception in exoplayer.
+            # EXT-X-START is not supposed to be within 3 target durations of the end,
+            # but this seems ok
+            playlist.append(
+                f"#EXT-X-START:TIME-OFFSET=-{1.5 * target_duration:.3f},PRECISE=YES"
```

Comment:
This if only when not using LL-HLS. The previous method did not specify a `EXT-X-START` tag so most of the players started at the third segment back from the live edge (this is actually the minimum specified by the spec and also the maximum that we include in our playlists). However it seems like players can actually violate this spec, and starting from 1.5 (which should get rounded to 2) segments back doesn't seem to negatively impact things while reducing latency by one segment.